src/hlsBinaries.ts

```diff
@@ -79,364 +51,310 @@ class MissingToolError extends Error {
+    logger.info('Downloading ghcup');
 
-  const githubOS = getGithubOS();
-  if (githubOS === null) {
-    // Don't have any binaries available for this platform
-    window.showErrorMessage(`Couldn't find any pre-built haskell-language-server binaries for ${process.platform}`);
-    return null;
-  }
+    const storagePath: string = await getStoragePath(context);
+    logger.info(`Using ${storagePath} to store downloaded binaries`);
 
-  logger.info('Fetching the latest release from GitHub or from cache');
-  const releases = await getReleaseMetadata(context, storagePath, logger);
-  const updateBehaviour = workspace.getConfiguration('haskell').get('updateBehavior') as UpdateBehaviour;
-  if (!releases) {
-    let message = "Couldn't find any pre-built haskell-language-server binaries";
-    if (updateBehaviour === 'never-check') {
-      message += ' (and checking for newer versions is disabled)';
+    if (!fs.existsSync(storagePath)) {
+        fs.mkdirSync(storagePath);
     }
-    window.showErrorMessage(message);
-    return null;
-  }
-  logger.info(`The latest known release is ${releases[0].tag_name}`);
-  logger.info('Figure out the ghc version to use or advertise an installation link for missing components');
-  const dir: string = folder?.uri?.fsPath ?? path.dirname(resource.fsPath);
-  let ghcVersion: string;
-  try {
-    ghcVersion = await getProjectGhcVersion(context, logger, dir, releases[0], storagePath);
-  } catch (error) {
-    if (error instanceof MissingToolError) {
-      const link = error.installLink();
-      if (link) {
-        if (await window.showErrorMessage(error.message, `Install ${error.tool}`)) {
-          env.openExternal(link);
-        }
-      } else {
-        await window.showErrorMessage(error.message);
-      }
-    } else if (error instanceof NoBinariesError) {
-      window.showInformationMessage(error.message);
-    } else if (error instanceof Error) {
-      // We couldn't figure out the right ghc version to download
-      window.showErrorMessage(`Couldn't figure out what GHC version the project is using: ${error.message}`);
-    }
-    return null;
-  }
 
-  // When searching for binaries, use startsWith because the compression may differ
-  // between .zip and .gz
-  const assetName = `haskell-language-server-${githubOS}-${ghcVersion}${exeExt}`;
-  logger.info(`Search for binary ${assetName} in release assets`);
-  const release = releases?.find(r => r.assets.find((x) => x.name.startsWith(assetName)));
-  const asset = release?.assets.find((x) => x.name.startsWith(assetName));
-  if (!asset) {
-    let msg = new NoBinariesError(releases[0].tag_name, ghcVersion).message;
-    if (updateBehaviour === 'never-check') {
-      msg += ". Consider set 'haskell.updateBehaviour' to 'up-to-date' to check if another release includes the missing binary";
+    const ghcup = path.join(storagePath, 'ghcup');
```

Comment:
We could also check if the user already has GHCup and use that if it exists.

Reply:
But then we need to upgrade it before using, etc.

Reply:
Yes, I want to avoid messing with user-installed ghcup.